File: processing/src/processing/YUV.java
package processing;

/**
 * Color stored in YUV space.
 * 
 * Colors are stored as values from:
 * y -> [0 - 1]
 * u -> [-0.5 - 0.5]
 * v -> [-0.5 - 0.5]
 * 
 * @author Christopher Gebhardt
 * @date 04.06.2014
 *
 */
public class YUV {

	public float y = 0;
	public float u = 0;
	public float v = 0;

	/**
	 * conversion matrix for rgb to yuv.
	 * [conversionMatrix] * (rgb)
	 */
	float[] convMatrix = new float[]{
			0.299f, 0.587f, 0.114f,
			-0.14713f, -0.28886f, 0.436f,
			0.615f, -0.51499f, -0.10001f
	};

	public YUV(RGB rgb){
		y = convMatrix[0] * rgb.r + convMatrix[1] * rgb.g + convMatrix[2] * rgb.b;
		u = convMatrix[3] * rgb.r + convMatrix[4] * rgb.g + convMatrix[5] * rgb.b;
		v = convMatrix[6] * rgb.r + convMatrix[7] * rgb.g + convMatrix[8] * rgb.b;
	}
	
	public YUV(float y, float u, float v, boolean isYuv){
		this.y = y;
		this.u = u;
		this.v = v;
	}

	public YUV(String hexString){
		if(hexString.length() == 6){
			float r = Integer.parseInt(hexString.substring(0,2), 16) / 255f;
			float g = Integer.parseInt(hexString.substring(2,4), 16) / 255f;
			float b = Integer.parseInt(hexString.substring(4), 16) / 255f;

			y = convMatrix[0] * r + convMatrix[1] * g + convMatrix[2] * b;
			u = convMatrix[3] * r + convMatrix[4] * g + convMatrix[5] * b;
			v = convMatrix[6] * r + convMatrix[7] * g + convMatrix[8] * b;
		} else {
			System.out.println("RGB: not a valid hex value");
		}
	}

	public YUV(int red, int green, int blue){
		float r = (float)red / 255f;
		float g = (float)green / 255f;
		float b = (float)blue / 255f;

		y = convMatrix[0] * r + convMatrix[1] * g + convMatrix[2] * b;
		u = convMatrix[3] * r + convMatrix[4] * g + convMatrix[5] * b;
		v = convMatrix[6] * r + convMatrix[7] * g + convMatrix[8] * b;
	}

	public YUV(float red, float green, float blue){
		y = convMatrix[0] * red + convMatrix[1] * green + convMatrix[2] * blue;
		u = convMatrix[3] * red + convMatrix[4] * green + convMatrix[5] * blue;
		v = convMatrix[6] * red + convMatrix[7] * green + convMatrix[8] * blue;
	}

	/**
	 * calculate distance between two colors
	 * @param yuv
	 * @return
	 */
	public float distanceTo(YUV yuv){
//		return (float)Math.sqrt((u-yuv.u)*(u-yuv.u) + (v-yuv.v)*(v-yuv.v));
		return (float)Math.sqrt((u-yuv.u)*(u-yuv.u) + (v-yuv.v)*(v-yuv.v) + (y-yuv.y)*(y-yuv.y));
	}
	
	/**
	 * Interpolate this color with another color
	 * @param yuv
	 * @return
	 */
	public YUV average(YUV yuv){
		float ny = (y + yuv.y) / 2f;
		float nu = (u + yuv.u) / 2f;
		float nv = (v + yuv.v) / 2f;
		
		return new YUV(ny, nu, nv, true);
	}

	public void set(RGB rgb){
		y = convMatrix[0] * rgb.r + convMatrix[1] * rgb.g + convMatrix[2] * rgb.b;
		u = convMatrix[3] * rgb.r + convMatrix[4] * rgb.g + convMatrix[5] * rgb.b;
		v = convMatrix[6] * rgb.r + convMatrix[7] * rgb.g + convMatrix[8] * rgb.b;
	}
	
	public void set(float y, float u, float v, boolean isYuv){
		this.y = y;
		this.u = u;
		this.v = v;
	}
	
	public RGB getRGB(){
		return new RGB(this);
	}

	public boolean equals(YUV yuv){
		if(yuv == null){
			return false;
		}

		if(y == yuv.y && u == yuv.u && v == yuv.v){
			return true;
		} else {
			return false;
		}
	}

	public String toString(){
		return "["+y+","+u+","+v+"]";
	}
}

File: python/recipe-printer-server/YUV.py
#!/usr/bin/env python2

# YUV color class
# colors are stored as yuv values in range
# y = [0 - 1]
# u = [-0.5 - 0.5]
# v = [-0.5 - 0.5]

import math
from RGB import *

class YUV(object):
    y = 0
    u = 0
    v = 0
    
    convMatrix = [
        0.299, 0.587, 0.114,
        -0.14713, -0.28886, 0.436,
        0.615, -0.51499, -0.10001]
    
    def __init__(self, y=0, u=0, v=0):
        self.y = y
        self.u = u
        self.v = v
        
    @classmethod
    def fromIntRgb(cls, r=0, g=0, b=0):
        red = r / 255.0;
        green = g / 255.0;
        blue = b / 255.0;
        
        y = cls.convMatrix[0] * red + cls.convMatrix[1] * green + cls.convMatrix[2] * blue
        u = cls.convMatrix[3] * red + cls.convMatrix[4] * green + cls.convMatrix[5] * blue
        v = cls.convMatrix[6] * red + cls.convMatrix[7] * green + cls.convMatrix[8] * blue
        return cls(y, u, v)
    
    @classmethod
    def fromFloatRgb(cls, r=0, g=0, b=0):
        y = cls.convMatrix[0] * r + cls.convMatrix[1] * g + cls.convMatrix[2] * b
        u = cls.convMatrix[3] * r + cls.convMatrix[4] * g + cls.convMatrix[5] * b
        v = cls.convMatrix[6] * r + cls.convMatrix[7] * g + cls.convMatrix[8] * b
        return cls(y, u, v)
    
    @classmethod
    def fromYUV(cls, yuv=None):
        if yuv == None:
            return cls()
        else:
            return cls(yuv.y, yuv.u, yuv.v)
        
    @classmethod
    def fromRGB(cls, rgb=None):
        if yuv == None:
            return cls()
        else:
            y = cls.convMatrix[0] * rgb.r + cls.convMatrix[1] * rgb.g + cls.convMatrix[2] * rgb.b
            u = cls.convMatrix[3] * rgb.r + cls.convMatrix[4] * rgb.g + cls.convMatrix[5] * rgb.b
            v = cls.convMatrix[6] * rgb.r + cls.convMatrix[7] * rgb.g + cls.convMatrix[8] * rgb.b
            return cls(y, u, v)
    
    #def setcolorHEX(self, hexString):
    #    if len(hexString) == 6:
    #        r = int(hexString[0:2], 16) / 255.0
    #        g = Integer.parseInt(hexString[2,4], 16) / 255.0
    #        b = Integer.parseInt(hexString[4:], 16) / 255.0
    #
    #        self.y = self.convMatrix[0] * r + self.convMatrix[1] * g + self.convMatrix[2] * b
    #        self.u = self.convMatrix[3] * r + self.convMatrix[4] * g + self.convMatrix[5] * b
    #        self.v = self.convMatrix[6] * r + self.convMatrix[7] * g + self.convMatrix[8] * b
    #    else:
    #        print("RGB: not a valid hex value")
     
    def distanceTo(self, yuv):
        return math.sqrt((self.u-yuv.u)*(self.u-yuv.u) + (self.v-yuv.v)*(self.v-yuv.v) + (self.y-yuv.y)*(self.y-yuv.y))
    
    def average(self, yuv):
        ny = (self.y + yuv.y) / 2.0;
        nu = (self.u + yuv.u) / 2.0;
        nv = (self.v + yuv.v) / 2.0;
        
        return YUV(ny, nu, nv)
    
    def equals(self, yuv):
        if yuv == None:
            return false
        
        if self.y == yuv.y and self.u == yuv.u and self.v == yuv.v:
            return True
        else:
            return False
        
    def toString(self):
        return "["+str(self.y)+","+str(self.u)+","+str(self.v)+"]"
    
    def copy(self):
        return YUV(self.y, self.u, self.v)
    
    def getRGB(self):
        return RGB.fromFloatYUV(self.y, self.u, self.v)
        
